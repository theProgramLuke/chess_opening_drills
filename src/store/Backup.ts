import _ from "lodash";
import fs from "graceful-fs";
import path from "path";

export const AgeSeparator = "-";

export class Backup {
  filePath: string;
  age: number;

  constructor(filePath: string) {
    this.filePath = filePath;
    this.age = _.parseInt(this.filePath.split(AgeSeparator)[1]);
  }

  async delete(): Promise<void> {
    await fs.unlink(this.filePath, _.noop);
  }

  async save(content: string): Promise<void> {
    await fs.mkdir(
      path.dirname(this.filePath),
      { recursive: true },
      async () => {
        await fs.writeFile(this.filePath, content, _.noop);
      }
    );
  }
}
