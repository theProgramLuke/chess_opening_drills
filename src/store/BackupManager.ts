import fs from "graceful-fs";
import path from "path";
import _ from "lodash";

import { Backup } from "@/store/Backup";

export const day = 24 * 60 * 60 * 1000;
export const month = day * 31;
export const year = day * 365;

const lastBackupAge = (backups: Backup[]): number => {
  return _.min(_.map(backups, backup => backup.age)) || 0;
};

const trimBackups = (backups: Backup[], limit: number): Backup[] => {
  const sorted = _.sortBy(backups, backup => 0 - backup.age);

  while (sorted.length > limit) {
    const toDelete = sorted.pop();
    if (toDelete) {
      toDelete.delete();
    }
  }

  return sorted;
};

export class BackupManager {
  backupFolder: string;
  dailyBackups: Backup[];
  monthlyBackups: Backup[];
  yearlyBackups: Backup[];
  dailyLimit: number;
  monthlyLimit: number;
  yearlyLimit: number;
  createBackup: (filePath: string) => Backup;

  constructor(
    backupFolder: string,
    dailyLimit: number,
    monthlyLimit: number,
    yearlyLimit: number,
    listDirectory = (directory: string): string[] => fs.readdirSync(directory),
    createBackup = (filePath: string) => new Backup(filePath)
  ) {
    this.backupFolder = backupFolder;
    this.createBackup = createBackup;
    this.dailyBackups = this.discoverBackups(
      "daily",
      listDirectory,
      createBackup
    );
    this.monthlyBackups = this.discoverBackups(
      "monthly",
      listDirectory,
      createBackup
    );
    this.yearlyBackups = this.discoverBackups(
      "yearly",
      listDirectory,
      createBackup
    );
    this.dailyLimit = dailyLimit;
    this.monthlyLimit = monthlyLimit;
    this.yearlyLimit = yearlyLimit;
  }

  private discoverBackups(
    frequency: string,
    listDirectory: (directory: string) => string[],
    createBackup: (filePath: string) => Backup
  ): Backup[] {
    try {
      const files = listDirectory(path.join(this.backupFolder, frequency));
      return _.map(files, createBackup);
    } catch {
      return [];
    }
  }

  SaveBackup(content: string, now = _.now): void {
    const fileName = `settings-${now()}.json`;
    let backup: Backup;

    if (this.dailyLimit && now() - lastBackupAge(this.dailyBackups) >= day) {
      backup = this.createBackup(
        path.join(this.backupFolder, "daily", fileName)
      );
      this.dailyBackups.push(backup);
      backup.save(content);
    }
    this.dailyBackups = trimBackups(this.dailyBackups, this.dailyLimit);

    if (
      this.monthlyLimit &&
      now() - lastBackupAge(this.monthlyBackups) >= month
    ) {
      backup = this.createBackup(
        path.join(this.backupFolder, "monthly", fileName)
      );
      this.monthlyBackups.push(backup);
      backup.save(content);
    }
    this.monthlyBackups = trimBackups(this.monthlyBackups, this.monthlyLimit);

    if (this.yearlyLimit && now() - lastBackupAge(this.yearlyBackups) >= year) {
      backup = this.createBackup(
        path.join(this.backupFolder, "yearly", fileName)
      );
      this.yearlyBackups.push(backup);
      backup.save(content);
    }
    this.yearlyBackups = trimBackups(this.yearlyBackups, this.yearlyLimit);
  }
}
