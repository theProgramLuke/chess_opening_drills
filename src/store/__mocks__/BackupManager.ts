import _ from "lodash";

import { Backup } from "@/store/Backup";

export const day = 24 * 60 * 60 * 1000;
export const month = day * 31;
export const year = day * 365;

export class BackupManager {
  backupFolder: string;
  dailyBackups: Backup[];
  monthlyBackups: Backup[];
  yearlyBackups: Backup[];
  dailyLimit: number;
  monthlyLimit: number;
  yearlyLimit: number;

  constructor(
    backupFolder: string,
    dailyLimit: number,
    monthlyLimit: number,
    yearlyLimit: number
  ) {
    this.backupFolder = backupFolder;
    this.dailyBackups = [];
    this.monthlyBackups = [];
    this.yearlyBackups = [];
    this.dailyLimit = dailyLimit;
    this.monthlyLimit = monthlyLimit;
    this.yearlyLimit = yearlyLimit;
    this.SaveBackup = jest.fn();
  }

  SaveBackup: (getContent: () => string, now?: () => number) => void;
}
