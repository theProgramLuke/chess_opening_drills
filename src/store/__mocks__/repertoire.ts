import _ from "lodash";

import {
  RepertoirePosition,
  SavedRepertoirePosition
} from "@/store/repertoirePosition";
import { RepertoireTag, SavedRepertoireTag } from "@/store/repertoireTag";
import { Move } from "@/store/move";
import { PgnGame } from "pgn-parser";

export class SavedRepertoire {
  positions: SavedRepertoirePosition[];
  tags: SavedRepertoireTag[];

  constructor(
    positions: SavedRepertoirePosition[],
    tags: SavedRepertoireTag[]
  ) {
    this.positions = positions;
    this.tags = tags;
  }
}

export class Repertoire {
  positions: RepertoirePosition[];
  tags: RepertoireTag[];

  constructor(positions: RepertoirePosition[], tags: RepertoireTag[]) {
    this.positions = positions;
    this.tags = tags;
    this.AddMove = jest.fn();
    this.RemoveRepertoireTag = jest.fn();
    this.RemoveMove = jest.fn();
    this.AddFromGame = jest.fn();
    this.AsSaved = jest.fn();
  }

  AddMove: (parent: RepertoirePosition, move: Move) => void;

  RemoveRepertoireTag: (tagToRemove: RepertoireTag) => void;

  RemoveMove: (move: Move) => void;

  AddFromGame: (game: PgnGame) => void;

  AsSaved: () => SavedRepertoire;

  static FromSaved: (saved: SavedRepertoire) => Repertoire = jest.fn();
}
