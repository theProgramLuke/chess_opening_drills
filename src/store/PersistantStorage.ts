import ElectronStore from "electron-store";
import _ from "lodash";
import fs from "graceful-fs";
import path from "path";

import { RepertoireTag } from "./repertoireTag";
import { Side } from "./side";
import { RepertoirePosition } from "@/store/repertoirePosition";
import { Repertoire, SavedRepertoire } from "./repertoire";
import { EngineMetadata } from "./EngineHelpers";
import { BackupManager } from "./BackupManager";
import electron from "electron";

export interface SavedStorage {
  darkMode: boolean;
  primary: string;
  secondary: string;
  accent: string;
  error: string;
  warning: string;
  info: string;
  success: string;
  boardTheme: string;
  pieceTheme: string;
  whiteRepertoire: SavedRepertoire;
  blackRepertoire: SavedRepertoire;
  engineMetadata?: EngineMetadata;
  backupDirectory?: string;
  dailyBackupLimit: number;
  monthlyBackupLimit: number;
  yearlyBackupLimit: number;
}

export interface Storage {
  darkMode: boolean;
  primary: string;
  secondary: string;
  accent: string;
  error: string;
  warning: string;
  info: string;
  success: string;
  boardTheme: string;
  pieceTheme: string;
  whiteRepertoire: Repertoire;
  blackRepertoire: Repertoire;
  engineMetadata?: EngineMetadata;
  backupDirectory?: string;
  dailyBackupLimit: number;
  monthlyBackupLimit: number;
  yearlyBackupLimit: number;
}

function GetDefaultStorage() {
  const whiteStartPosition = new RepertoirePosition(
    "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
    "",
    Side.White,
    true
  );

  const blackStartPosition = new RepertoirePosition(
    "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
    "",
    Side.Black,
    false
  );

  const defaultCwd = (electron.app || electron.remote.app).getPath("userData");

  return new ElectronStore<SavedStorage>({
    defaults: {
      darkMode: false,
      primary: "#2196F3",
      secondary: "#424242",
      accent: "#FF4081",
      error: "#FF5252",
      warning: "#FFC107",
      info: "#2196F3",
      success: "#4CAF50",
      boardTheme: "maple",
      pieceTheme: "staunty",
      whiteRepertoire: new Repertoire(
        [whiteStartPosition],
        [
          new RepertoireTag(
            Side.White,
            "White",
            whiteStartPosition,
            whiteStartPosition.fen,
            [],
            "whiteStart"
          )
        ]
      ).AsSaved(),
      blackRepertoire: new Repertoire(
        [blackStartPosition],
        [
          new RepertoireTag(
            Side.Black,
            "Black",
            blackStartPosition,
            blackStartPosition.fen,
            [],
            "blackStart"
          )
        ]
      ).AsSaved(),
      backupDirectory: path.join(defaultCwd, "backups"),
      dailyBackupLimit: 14,
      monthlyBackupLimit: 6,
      yearlyBackupLimit: 3
    }
  });
}

export class PersistantStorage implements Storage {
  storage: ElectronStore<SavedStorage>;
  backupManager?: BackupManager;
  createBackupManager: (
    filePath: string,
    dailyBackupLimit: number,
    monthlyBackupLimit: number,
    yearlyBackupLimit: number
  ) => BackupManager;

  constructor(
    storage?: ElectronStore<SavedStorage>,
    createBackupManager = (
      filePath: string,
      dailyBackupLimit: number,
      monthlyBackupLimit: number,
      yearlyBackupLimit: number
    ) =>
      new BackupManager(
        filePath,
        dailyBackupLimit,
        monthlyBackupLimit,
        yearlyBackupLimit
      )
  ) {
    this.storage = storage || GetDefaultStorage();
    this.createBackupManager = createBackupManager;
    this.initializeBackupManagement();
  }

  private initializeBackupManagement(): void {
    if (this.backupDirectory) {
      this.backupManager = this.createBackupManager(
        this.backupDirectory,
        this.dailyBackupLimit,
        this.monthlyBackupLimit,
        this.yearlyBackupLimit
      );
    }
  }

  backup(): void {
    if (this.backupManager) {
      this.backupManager.SaveBackup(this.serialize());
    }
  }

  setStorage<Key extends keyof SavedStorage>(
    key: Key,
    value?: SavedStorage[Key]
  ): void {
    this.storage.set(key, value);
    this.backup();
  }

  get darkMode(): boolean {
    return this.storage.get("darkMode");
  }

  set darkMode(value: boolean) {
    this.setStorage("darkMode", value);
  }

  get primary(): string {
    return this.storage.get("primary");
  }

  set primary(value: string) {
    this.setStorage("primary", value);
  }

  get secondary(): string {
    return this.storage.get("secondary");
  }

  set secondary(value: string) {
    this.setStorage("secondary", value);
  }

  get accent(): string {
    return this.storage.get("accent");
  }

  set accent(value: string) {
    this.setStorage("accent", value);
  }

  get error(): string {
    return this.storage.get("error");
  }

  set error(value: string) {
    this.setStorage("error", value);
  }

  get warning(): string {
    return this.storage.get("warning");
  }

  set warning(value: string) {
    this.setStorage("warning", value);
  }

  get info(): string {
    return this.storage.get("info");
  }

  set info(value: string) {
    this.setStorage("info", value);
  }

  get success(): string {
    return this.storage.get("success");
  }

  set success(value: string) {
    this.setStorage("success", value);
  }

  get boardTheme(): string {
    return this.storage.get("boardTheme");
  }

  set boardTheme(value: string) {
    this.setStorage("boardTheme", value);
  }

  get pieceTheme(): string {
    return this.storage.get("pieceTheme");
  }

  set pieceTheme(value: string) {
    this.setStorage("pieceTheme", value);
  }

  get whiteRepertoire(): Repertoire {
    return Repertoire.FromSaved(this.storage.get("whiteRepertoire"));
  }

  set whiteRepertoire(value: Repertoire) {
    this.setStorage("whiteRepertoire", value.AsSaved());
  }

  get blackRepertoire(): Repertoire {
    return Repertoire.FromSaved(this.storage.get("blackRepertoire"));
  }

  set blackRepertoire(value: Repertoire) {
    this.setStorage("blackRepertoire", value.AsSaved());
  }

  get engineMetadata(): EngineMetadata | undefined {
    return this.storage.get("engineMetadata");
  }

  set engineMetadata(engineMetadata: EngineMetadata | undefined) {
    if (engineMetadata) {
      this.setStorage("engineMetadata", engineMetadata);
    } else {
      this.storage.delete("engineMetadata");
    }
  }

  get backupDirectory(): string | undefined {
    return this.storage.get("backupDirectory");
  }

  set backupDirectory(backupDirectory: string | undefined) {
    this.backupManager = undefined;
    if (backupDirectory) {
      this.setStorage("backupDirectory", backupDirectory);
      this.initializeBackupManagement();
    } else {
      this.storage.delete("backupDirectory");
    }
  }

  get dailyBackupLimit(): number {
    return this.storage.get("dailyBackupLimit");
  }

  set dailyBackupLimit(limit: number) {
    this.setStorage("dailyBackupLimit", limit);
    if (this.backupManager) {
      this.backupManager.dailyLimit = limit;
    }
  }

  get monthlyBackupLimit(): number {
    return this.storage.get("monthlyBackupLimit");
  }

  set monthlyBackupLimit(limit: number) {
    this.setStorage("monthlyBackupLimit", limit);
    if (this.backupManager) {
      this.backupManager.monthlyLimit = limit;
    }
  }

  get yearlyBackupLimit(): number {
    return this.storage.get("yearlyBackupLimit");
  }

  set yearlyBackupLimit(limit: number) {
    this.setStorage("yearlyBackupLimit", limit);
    if (this.backupManager) {
      this.backupManager.yearlyLimit = limit;
    }
  }

  serialize(
    readFile = (filePath: string): string =>
      fs.readFileSync(filePath, { encoding: "utf8" })
  ): string {
    return readFile(this.storage.path);
  }

  clear(): void {
    this.storage.clear();
  }
}
