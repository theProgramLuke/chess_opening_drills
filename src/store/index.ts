import Vue from "vue";
import Vuex from "vuex";

import { RepertoirePosition } from "@/store/repertoirePosition";
import { Move } from "./move";
import { GetPersistantStorage } from "./storage";
import { Repertoire } from "./repertoire";
import { RepertoireTag } from "./repertoireTag";
import { Side } from "./side";
import { TrainingEvent } from "./TrainingEvent";
import { PgnGame } from "./pgnParser";

Vue.use(Vuex);

const storage = GetPersistantStorage();

export default new Vuex.Store({
  state: {
    darkMode: storage.get("darkMode"),
    primary: storage.get("primary"),
    secondary: storage.get("secondary"),
    accent: storage.get("accent"),
    error: storage.get("error"),
    warning: storage.get("warning"),
    info: storage.get("info"),
    success: storage.get("success"),
    boardTheme: storage.get("boardTheme"),
    pieceTheme: storage.get("pieceTheme"),
    whiteRepertoire: Repertoire.FromSaved(storage.get("whiteRepertoire")),
    blackRepertoire: Repertoire.FromSaved(storage.get("blackRepertoire"))
  },

  mutations: {
    setDarkMode: (state, darkMode): void => {
      state.darkMode = darkMode;
      storage.set("darkMode", darkMode);
    },

    setBoardTheme: (state, boardTheme): void => {
      state.boardTheme = boardTheme;
      storage.set("boardTheme", boardTheme);
    },

    setPieceTheme: (state, pieceTheme): void => {
      state.pieceTheme = pieceTheme;
      storage.set("pieceTheme", pieceTheme);
    },

    setColor: function(
      state,
      payload: {
        colorToSet:
          | "primary"
          | "secondary"
          | "accent"
          | "error"
          | "warning"
          | "info"
          | "success";
        color: string;
      }
    ): void {
      state[payload.colorToSet] = payload.color;
      storage.set(payload.colorToSet, payload.color);
    },

    addRepertoirePosition: (
      state,
      payload: { parent: RepertoirePosition; newMove: Move }
    ): void => {
      if (payload.parent && payload.newMove) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddMove(payload.parent, payload.newMove);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addRepertoireTag: (
      state,
      payload: { parent: RepertoireTag; tag: RepertoireTag }
    ): void => {
      if (payload.parent && payload.tag) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.parent.AddChild(payload.tag);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addTrainingEvent: (
      state,
      payload: { position: RepertoirePosition; event: TrainingEvent }
    ): void => {
      if (payload.position && payload.event) {
        const repertoire =
          payload.position.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.position.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.position.AddTrainingEvent(payload.event);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    removeRepertoireTag: (state, tag: RepertoireTag): void => {
      const repertoire =
        tag.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        tag.forSide === Side.White ? "whiteRepertoire" : "blackRepertoire";

      repertoire.RemoveRepertoireTag(tag);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    removeRepertoireMove: (state, move: Move): void => {
      const repertoire =
        move.position.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        move.position.forSide === Side.White
          ? "whiteRepertoire"
          : "blackRepertoire";

      repertoire.RemoveMove(move);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    addPositionsFromGame(
      state,
      payload: { forSide: Side; game: PgnGame }
    ): void {
      if (payload.game) {
        const repertoire =
          payload.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddFromGame(payload.game);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    clearStorage(): void {
      storage.clear();
    }
  },

  actions: {},

  modules: {},

  strict: process.env.NODE_ENV !== "production"
});
