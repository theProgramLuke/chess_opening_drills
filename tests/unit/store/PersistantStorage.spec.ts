import ElectronStore from "electron-store";
import _ from "lodash";

import { SavedStorage, PersistantStorage } from "@/store/PersistantStorage";
import { Repertoire, SavedRepertoire } from "@/store/repertoire";
import { BackupManager } from "@/store/BackupManager";

jest.mock("electron-store");
jest.mock("@/store/repertoire");
jest.mock("@/store/BackupManager");

type Writeable<T> = { -readonly [P in keyof T]: T[P] };

describe("PersistantStorage", () => {
  const storagePath = "some/path.json";
  const createMockBackupManager = (
    filePath: string,
    dailyBackupLimit: number,
    monthlyBackupLimit: number,
    yearlyBackupLimit: number
  ) => {
    const manager = new BackupManager(
      filePath,
      dailyBackupLimit,
      monthlyBackupLimit,
      yearlyBackupLimit
    );
    manager.backupFolder = filePath;
    return manager;
  };

  let store: ElectronStore<SavedStorage>;
  let persistantStorage: PersistantStorage;

  beforeEach(() => {
    store = new ElectronStore<SavedStorage>();
    (store as Writeable<ElectronStore<SavedStorage>>).path = storagePath;
    persistantStorage = new PersistantStorage(store, createMockBackupManager);
  });

  describe("dark mode", () => {
    it.each([true, false])("should get the stored dark mode %s", darkMode => {
      store.get = jest.fn(() => darkMode);

      const actual = persistantStorage.darkMode;

      expect(store.get).toBeCalledWith("darkMode");
      expect(actual).toBe(darkMode);
    });

    it.each([true, false])("should set the stored dark mode %s", darkMode => {
      persistantStorage.darkMode = darkMode;

      expect(store.set).toBeCalledWith("darkMode", darkMode);
    });
  });

  describe("primary", () => {
    it.each(["red", "blue"])(
      "should get the stored primary color %s",
      color => {
        store.get = jest.fn(() => color);

        const actual = persistantStorage.primary;

        expect(store.get).toBeCalledWith("primary");
        expect(actual).toBe(color);
      }
    );

    it.each(["red", "blue"])(
      "should set the stored primary color %s",
      color => {
        persistantStorage.primary = color;

        expect(store.set).toBeCalledWith("primary", color);
      }
    );
  });

  describe("secondary", () => {
    it.each(["red", "blue"])(
      "should get the stored secondary color %s",
      color => {
        store.get = jest.fn(() => color);

        const actual = persistantStorage.secondary;

        expect(store.get).toBeCalledWith("secondary");
        expect(actual).toBe(color);
      }
    );

    it.each(["red", "blue"])(
      "should set the stored secondary color %s",
      color => {
        persistantStorage.secondary = color;

        expect(store.set).toBeCalledWith("secondary", color);
      }
    );
  });

  describe("accent", () => {
    it.each(["red", "blue"])("should get the stored accent color %s", color => {
      store.get = jest.fn(() => color);

      const actual = persistantStorage.accent;

      expect(store.get).toBeCalledWith("accent");
      expect(actual).toBe(color);
    });

    it.each(["red", "blue"])("should set the stored accent color %s", color => {
      persistantStorage.accent = color;

      expect(store.set).toBeCalledWith("accent", color);
    });
  });

  describe("error", () => {
    it.each(["red", "blue"])("should get the stored error color %s", color => {
      store.get = jest.fn(() => color);

      const actual = persistantStorage.error;

      expect(store.get).toBeCalledWith("error");
      expect(actual).toBe(color);
    });

    it.each(["red", "blue"])("should set the stored error color %s", color => {
      persistantStorage.error = color;

      expect(store.set).toBeCalledWith("error", color);
    });
  });

  describe("warning", () => {
    it.each(["red", "blue"])(
      "should get the stored warning color %s",
      color => {
        store.get = jest.fn(() => color);

        const actual = persistantStorage.warning;

        expect(store.get).toBeCalledWith("warning");
        expect(actual).toBe(color);
      }
    );

    it.each(["red", "blue"])(
      "should set the stored warning color %s",
      color => {
        persistantStorage.warning = color;

        expect(store.set).toBeCalledWith("warning", color);
      }
    );
  });

  describe("info", () => {
    it.each(["red", "blue"])("should get the stored info color %s", color => {
      store.get = jest.fn(() => color);

      const actual = persistantStorage.info;

      expect(store.get).toBeCalledWith("info");
      expect(actual).toBe(color);
    });

    it.each(["red", "blue"])("should set the stored info color %s", color => {
      persistantStorage.info = color;

      expect(store.set).toBeCalledWith("info", color);
    });
  });

  describe("success", () => {
    it.each(["red", "blue"])(
      "should get the stored success color %s",
      color => {
        store.get = jest.fn(() => color);

        const actual = persistantStorage.success;

        expect(store.get).toBeCalledWith("success");
        expect(actual).toBe(color);
      }
    );

    it.each(["red", "blue"])(
      "should set the stored success color %s",
      color => {
        persistantStorage.success = color;

        expect(store.set).toBeCalledWith("success", color);
      }
    );
  });

  describe("boardTheme", () => {
    it.each(["red", "blue"])(
      "should get the stored boardTheme color %s",
      color => {
        store.get = jest.fn(() => color);

        const actual = persistantStorage.boardTheme;

        expect(store.get).toBeCalledWith("boardTheme");
        expect(actual).toBe(color);
      }
    );

    it.each(["red", "blue"])(
      "should set the stored boardTheme color %s",
      color => {
        persistantStorage.boardTheme = color;

        expect(store.set).toBeCalledWith("boardTheme", color);
      }
    );
  });

  describe("pieceTheme", () => {
    it.each(["red", "blue"])(
      "should get the stored pieceTheme color %s",
      color => {
        store.get = jest.fn(() => color);

        const actual = persistantStorage.pieceTheme;

        expect(store.get).toBeCalledWith("pieceTheme");
        expect(actual).toBe(color);
      }
    );

    it.each(["red", "blue"])(
      "should set the stored pieceTheme color %s",
      color => {
        persistantStorage.pieceTheme = color;

        expect(store.set).toBeCalledWith("pieceTheme", color);
      }
    );
  });

  describe("whiteRepertoire", () => {
    it("should get the saved repertoire FromSaved", () => {
      const repertoire = new Repertoire([], []);
      Repertoire.FromSaved = jest.fn(() => repertoire);

      const actual = persistantStorage.whiteRepertoire;

      expect(store.get).toBeCalledWith("whiteRepertoire");
      expect(actual).toBe(repertoire);
    });

    it("should set the repertoire AsSaved", () => {
      const savedRepertoire = new SavedRepertoire([], []);
      const repertoire = new Repertoire([], []);
      repertoire.AsSaved = jest.fn(() => savedRepertoire);

      persistantStorage.whiteRepertoire = repertoire;

      expect(store.set).toBeCalledWith("whiteRepertoire", savedRepertoire);
    });
  });

  describe("blackRepertoire", () => {
    it("should get the saved repertoire FromSaved", () => {
      const repertoire = new Repertoire([], []);
      Repertoire.FromSaved = jest.fn(() => repertoire);

      const actual = persistantStorage.blackRepertoire;

      expect(store.get).toBeCalledWith("blackRepertoire");
      expect(actual).toBe(repertoire);
    });

    it("should set the repertoire AsSaved", () => {
      const savedRepertoire = new SavedRepertoire([], []);
      const repertoire = new Repertoire([], []);
      repertoire.AsSaved = jest.fn(() => savedRepertoire);

      persistantStorage.blackRepertoire = repertoire;

      expect(store.set).toBeCalledWith("blackRepertoire", savedRepertoire);
    });
  });

  describe("backupDirectory", () => {
    it("should get the stored backup directory", () => {
      const backupDirectory = "backups";
      store.get = jest.fn(() => backupDirectory);

      const actual = persistantStorage.backupDirectory;

      expect(store.get).toBeCalledWith("backupDirectory");
      expect(actual).toBe(backupDirectory);
    });

    it("should set the stored backup directory", () => {
      const backupDirectory = "backups";

      persistantStorage.backupDirectory = backupDirectory;

      expect(store.set).toBeCalledWith("backupDirectory", backupDirectory);
    });

    it("should set the stored backupManager", () => {
      const backupDirectory = "backups";
      store.get = jest.fn(() => backupDirectory);

      persistantStorage.backupDirectory = backupDirectory;

      expect(persistantStorage.backupManager).toBeDefined();
      if (persistantStorage.backupManager) {
        expect(persistantStorage.backupManager.backupFolder).toBe(
          backupDirectory
        );
      }
    });

    it("should set the stored backup directory if undefined", () => {
      persistantStorage.backupDirectory = undefined;

      expect(store.delete).toBeCalledWith("backupDirectory");
    });

    it("should unset the backup manager if undefined", () => {
      persistantStorage.backupManager = new BackupManager("", 0, 0, 0);

      persistantStorage.backupDirectory = undefined;

      expect(persistantStorage.backupManager).toBeUndefined();
    });
  });

  describe("dailyBackupLimit", () => {
    it("should get the stored daily backup limit", () => {
      const limit = 2;
      store.get = jest.fn(() => limit);

      const actual = persistantStorage.dailyBackupLimit;

      expect(store.get).toBeCalledWith("dailyBackupLimit");
      expect(actual).toBe(limit);
    });

    it("should set the stored daily backup limit", () => {
      const limit = 2;
      persistantStorage.backupManager = new BackupManager("", 0, 0, 0);
      persistantStorage.serialize = jest.fn();

      persistantStorage.dailyBackupLimit = limit;

      expect(store.set).toBeCalledWith("dailyBackupLimit", limit);
      expect(persistantStorage.backupManager.dailyLimit).toBe(limit);
    });
  });

  describe("monthlyBackupLimit", () => {
    it("should get the stored monthly backup limit", () => {
      const limit = 2;
      store.get = jest.fn(() => limit);

      const actual = persistantStorage.monthlyBackupLimit;

      expect(store.get).toBeCalledWith("monthlyBackupLimit");
      expect(actual).toBe(limit);
    });

    it("should set the stored monthly backup limit", () => {
      const limit = 2;
      persistantStorage.backupManager = new BackupManager("", 0, 0, 0);
      persistantStorage.serialize = jest.fn();

      persistantStorage.monthlyBackupLimit = limit;

      expect(store.set).toBeCalledWith("monthlyBackupLimit", limit);
      expect(persistantStorage.backupManager.monthlyLimit).toBe(limit);
    });
  });

  describe("yearlyBackupLimit", () => {
    it("should get the stored yearly backup limit", () => {
      const limit = 2;
      store.get = jest.fn(() => limit);

      const actual = persistantStorage.yearlyBackupLimit;

      expect(store.get).toBeCalledWith("yearlyBackupLimit");
      expect(actual).toBe(limit);
    });

    it("should set the stored yearly backup limit", () => {
      const limit = 2;
      persistantStorage.backupManager = new BackupManager("", 0, 0, 0);
      persistantStorage.serialize = jest.fn();

      persistantStorage.yearlyBackupLimit = limit;

      expect(store.set).toBeCalledWith("yearlyBackupLimit", limit);
      expect(persistantStorage.backupManager.yearlyLimit).toBe(limit);
    });
  });

  describe("enableBackups", () => {
    it.each([true, false])(
      "should get the stored enable backups setting",
      enableBackups => {
        store.get = jest.fn(() => enableBackups);

        const actual = persistantStorage.enableBackups;

        expect(store.get).toBeCalledWith("enableBackups");
        expect(actual).toBe(enableBackups);
      }
    );

    it.each([true, false])(
      "should set the stored enable backups setting",
      enableBackups => {
        persistantStorage.enableBackups = enableBackups;

        expect(store.set).toBeCalledWith("enableBackups", enableBackups);
      }
    );
  });

  describe("serialize", () => {
    it("should get the persisted file content", () => {
      const content = "content";
      const readFileSync = jest.fn(() => content);

      const serialized = persistantStorage.serialize(readFileSync);

      expect(serialized).toBe(content);
    });
  });

  describe("backup", () => {
    it("should saveBackup with the serialized storage", async () => {
      store.get = () => true;
      const content = "content";
      const serialize = () => content;
      const backupManager = new BackupManager("", 0, 0, 0);
      persistantStorage.serialize = serialize;
      persistantStorage.backupManager = backupManager;
      let serializer = () => "";
      backupManager.SaveBackup = (fn: () => string) => (serializer = fn);

      await persistantStorage.backup();
      const actualContent = serializer();

      expect(actualContent).toEqual(content);
    });

    it("should not saveBackup when enable backups is false", async () => {
      store.get = () => false;
      const backupManager = new BackupManager("", 0, 0, 0);
      persistantStorage.backupManager = backupManager;

      await persistantStorage.backup();

      expect(backupManager.SaveBackup).not.toBeCalled();
    });

    it("should backup any changes", () => {
      const storage = new PersistantStorage(store, createMockBackupManager);
      storage.backup = jest.fn();

      storage.darkMode = true;

      expect(storage.backup).toBeCalled();
    });
  });

  describe("clear", () => {
    it("should clear the store", () => {
      persistantStorage.clear();

      expect(store.clear).toBeCalled();
    });
  });
});
