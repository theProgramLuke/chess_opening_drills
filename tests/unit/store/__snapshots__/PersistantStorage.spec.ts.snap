// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersistantStorage GetDefaultStorage should get sane starting values 1`] = `
Array [
  Array [],
  Array [
    Object {
      "defaults": Object {
        "accent": "#FF4081",
        "backupDirectory": "tmp/backups",
        "blackRepertoire": undefined,
        "boardTheme": "maple",
        "dailyBackupLimit": 14,
        "darkMode": false,
        "enableBackups": true,
        "error": "#FF5252",
        "info": "#2196F3",
        "monthlyBackupLimit": 6,
        "moveAnimationSpeed": 100,
        "pieceTheme": "staunty",
        "primary": "#2196F3",
        "secondary": "#424242",
        "success": "#4CAF50",
        "warning": "#FFC107",
        "whiteRepertoire": undefined,
        "yearlyBackupLimit": 25,
      },
    },
  ],
]
`;
